tm2/pkg/bft/consensus/state_test.go

```diff
@@ -183,13 +185,16 @@ func TestStateEnterProposeYesPrivValidator(t *testing.T) {
 		cs.Stop()
 		cs.Wait()
 	}()
+	defer ensureDrainedChannels(t, proposalCh, newRoundCh, timeoutCh)
 
 	// Wait for new round so proposer is set.
 	ensureNewRound(newRoundCh, height, round)
 	ensureNewProposal(proposalCh, height, round)
 
+	// XXX(FIXME): new round loop block get round state
+
 	// Check that Proposal, ProposalBlock, ProposalBlockParts are set.
-	rs := cs.GetRoundState()
+	rs := ensureGetRoundState(cs)
```

Comment:
looks like this is failing because ensureGetRoundState() waits, and meanwhile there was another round produced.
do we need the time.Sleep(5) in ensureGetRoundState(cs)? If so it's probably masking some bug, so we should just get rid of that. And then this function will complete before the next round. 

------
FUTURE WORK:

Even with the above, it's still technically racy because we're assuming that cs.GetRoundState() will complete before the next round starts. Even for the CI this should generally hold, but it's not guaranteed. The complete solution should be to allow the consensus reactor to be given allowances of execution, so that the test can give it more allowances (of rounds, or messages to process from the WAL). This is a bit more challenging but not that much more so. We basically need to support maxSteps and make sure only this many messages from either the peerMsgQueue or internalMsgQueue (I think both, but not 100% sure) are processed in ConsensusState.receiveRoutine, and then to exit that routine once the maxSteps allowance is complete. And then in the test we need to give it more allowance, which re-starts the receiveRoutine (in a separate goroutine) with the allowed maxSteps. If we are going to do this, we should convert one or few test cases first. 

-----

For now I'm pretty sure we can just remove time.Sleep(5) in ensureGetRoundState(). Do we really need ensureGetRoundState() separate from .GetRoundState()? e.g. does it actually help with debug/trace for the developer to know what's going on? If it doesn't help, we can also revert ensureGetRoundState() to .GetRoundState() as it is simpler.

Reply:
Thanks, @jaekwon. I wasn't sure if these tests should work without `ensureGetRoundState`. The sole purpose of `ensureGetRoundState` was to:
-  Delay the check of GetRoundState so that **the test fails locally**.
-  Add a timeout so that **the test fails early** (no need to wait 10 minutes).
The delay wasn't meant to be pushed in the first place, but it helped me test locally until all the issues were resolved. I was thinking that the tests should work with or without the delay as (as you says) it leave some tests technically racy.
But If you confirm this is sufficient for now, I am content with the current state. :)
I will cleanup the PR and it should be good for a final review

Reply:
Regarding the "FUTURE WORK" section, I will definitely take a look to see if it cannot be a quick win.